test/powershell/Language/Classes/Scripting.Classes.Attributes.Tests.ps1

```diff
@@ -216,3 +218,96 @@ Describe 'Type resolution with attributes' -Tag "CI" {
         }
     }
 }
+
+Describe 'ValidateSet support a dynamically generated set' -Tag "CI" {
+    Context 'C# test' {
+
+        BeforeAll {
+            $a=@'
+            using System;
+            using System.Management.Automation;
+            using System.Collections.Generic;
+
+            namespace Test.Language {
+
+                [Cmdlet(VerbsCommon.Get, "TestValidateSet1")]
+                public class TestValidateSetCommand1 : PSCmdlet
+                {
+                    [Parameter]
+                    [ValidateSet(typeof(GenValuesForParam1))]
+                    public string Param1;
+
+                    protected override void EndProcessing()
+                    {
+                        WriteObject(Param1);
+                    }
+                }
+
+                [Cmdlet(VerbsCommon.Get, "TestValidateSet2")]
+                public class TestValidateSetCommand2 : PSCmdlet
+                {
+                    [Parameter]
+                    [ValidateSet(typeof(PSCmdlet))]
+                    public string Param1;
+
+                    protected override void EndProcessing()
+                    {
+                        WriteObject(Param1);
+                    }
+                }
+
+                [Cmdlet(VerbsCommon.Get, "TestValidateSet3")]
+                public class TestValidateSetCommand3 : PSCmdlet
+                {
+                    [Parameter]
+                    [ValidateSet(typeof(GenValuesForParam3))]
+                    public string Param1;
+
+                    protected override void EndProcessing()
+                    {
+                        WriteObject(Param1);
+                    }
+                }
+
+
+                /// Implement of test IValidateSetValuesGenerator
+                public class GenValuesForParam1 : IValidateSetValuesGenerator
+                {
+                    public IEnumerable<string> GetValidValues()
+                    {
+                        var testValues = new string[] {"Test1","TestString","Test2"};
+                        foreach (var value in testValues)
+                        {
+                            yield return value;
+                        }
+                    }
+                }
+
+                /// Implement of test IValidateSetValuesGenerator to return Null

```

Comment:
I recommend learning more about how the `yield` keyword is implemented under the covers, e.g. this is a good article: http://csharpindepth.com/Articles/Chapter6/IteratorBlockImplementation.aspx